### PR TITLE
meta, structure: support snapshot meta.

### PR DESCRIPTION
**Pull request by coocood**
Snapshot meta will be used to load history schema, so we can
support history read with changed schema version.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala @hanfei1991 PTAL

File: meta/meta.go [hunk 1]
```diff
@@ -83,7 +83,13 @@ type Meta struct {
 
 // NewMeta creates a Meta in transaction txn.
 func NewMeta(txn kv.Transaction) *Meta {
-	t := structure.NewStructure(txn, []byte{'m'})
+	t := structure.NewStructure(txn, txn, []byte{'m'})
+	return &Meta{txn: t}
+}
+
+// NewSnapshotMeta create a Meta with snapshot.
```
Comment (hanfei1991) [inline meta/meta.go]: creates

File: structure/structure.go [hunk 2]
```diff
@@ -24,26 +24,30 @@ const (
...
 	errInvalidHashKeyFlag   = terror.ClassStructure.New(codeInvalidHashKeyFlag, "invalid encoded hash key flag")
 	errInvalidHashKeyPrefix = terror.ClassStructure.New(codeInvalidHashKeyPrefix, "invalid encoded hash key prefix")
 	errInvalidListIndex     = terror.ClassStructure.New(codeInvalidListMetaData, "invalid list index")
 	errInvalidListMetaData  = terror.ClassStructure.New(codeInvalidListMetaData, "invalid list meta data")
+	errWriteOnSnapshot      = terror.ClassStructure.New(codeWriteOnSnapshot, "write on snapshot")
 )
 
-// NewStructure creates a TxStructure in transaction txn and with key prefix.
-func NewStructure(txn kv.Transaction, prefix []byte) *TxStructure {
+// NewStructure creates a TxStructure and with Retriever, RetrieverMutator and key prefix.
```
Comment (hanfei1991) [inline structure/structure.go]: and with -> with ?

File: meta/meta.go [hunk 3]
```diff
@@ -83,7 +83,13 @@ type Meta struct {
 
 // NewMeta creates a Meta in transaction txn.
 func NewMeta(txn kv.Transaction) *Meta {
-	t := structure.NewStructure(txn, []byte{'m'})
+	t := structure.NewStructure(txn, txn, []byte{'m'})
```
Comment (hanfei1991) [inline meta/meta.go]: Can we set 'm' to a constant string?

File: structure/structure.go (hunk 2 above)
Comment (GregoryIan) [inline structure/structure.go]: that contains Retriever...?

File: structure/structure.go [hunk 4]
```diff
@@ -24,26 +24,30 @@ const (
...
+		remu:   remu,
 		prefix: prefix,
 	}
 }
 
 // TxStructure supports some simple data structures like string, hash, list, etc... and
 // you can use these in a transaction.
 type TxStructure struct {
-	txn    kv.Transaction
+	retr   kv.Retriever
```
Comment (shenli) [inline structure/structure.go]: How about rename them to reader and writer?

File: structure/structure.go [hunk 5]
```diff
@@ -24,26 +24,30 @@ const (
...
 	errInvalidHashKeyPrefix = terror.ClassStructure.New(codeInvalidHashKeyPrefix, "invalid encoded hash key prefix")
 	errInvalidListIndex     = terror.ClassStructure.New(codeInvalidListMetaData, "invalid list index")
 	errInvalidListMetaData  = terror.ClassStructure.New(codeInvalidListMetaData, "invalid list meta data")
+	errWriteOnSnapshot      = terror.ClassStructure.New(codeWriteOnSnapshot, "write on snapshot")
 )
 
-// NewStructure creates a TxStructure in transaction txn and with key prefix.
-func NewStructure(txn kv.Transaction, prefix []byte) *TxStructure {
+// NewStructure creates a TxStructure and with Retriever, RetrieverMutator and key prefix.
+func NewStructure(retr kv.Retriever, remu kv.RetrieverMutator, prefix []byte) *TxStructure {
```
Comment (shenli) [inline structure/structure.go]: Use Mutator as the second arg's type is enough?

**Comment by coocood** [comment]
@shenli @hanfei1991 PTAL

**Comment by zimulala** [comment]
LGTM

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**